src/net/epoll_event.cc

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2023-present, Qihoo, Inc.  All rights reserved.
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include "epoll_event.h"
+
+#ifdef HAVE_EPOLL
+
+#  include "callback_function.h"
+
+namespace net {
+
+const int BaseEvent::EVENT_READ = EPOLLIN;
+const int BaseEvent::EVENT_WRITE = EPOLLOUT;
+const int BaseEvent::EVENT_ERROR = EPOLLERR;
+const int BaseEvent::EVENT_HUB = EPOLLHUP;
+
+bool EpollEvent::Init() {
+  evFd_ = epoll_create1(0);
+  if (evFd_ == -1) {  // If the epoll creation fails, return false
+    return false;
+  }
+  if (mode_ & EVENT_MODE_READ) {  // Add the listen socket to epoll for read
+    AddEvent(listen_->Fd(), EVENT_READ | EVENT_ERROR | EVENT_HUB);
+  }
+  pipe(pipeFd);
+  AddEvent(pipeFd[0], EVENT_READ | EVENT_ERROR | EVENT_HUB);
+
+  return true;
+}
+
+void EpollEvent::AddEvent(int fd, int mask) {
+  struct epoll_event ev {};
+  ev.events = mask;
+  ev.data.fd = fd;
+  epoll_ctl(EvFd(), EPOLL_CTL_ADD, fd, &ev);
+}
+
+void EpollEvent::DelEvent(int fd) { epoll_ctl(EvFd(), EPOLL_CTL_DEL, fd, nullptr); }
+
+void EpollEvent::EventPoll() {
+  if (mode_ & EVENT_MODE_READ) {  // If it is a read multiplex, call EventRead
+    EventRead();
+  } else {  // If it is a write multiplex, call EventWrite
+    EventWrite();
+  }
+}
+
+void EpollEvent::AddWriteEvent(int fd) {
+  if (mode_ & EVENT_MODE_READ) {  // If it is a read multiplex, modify the event
+    struct epoll_event ev {};
+    ev.events = EVENT_WRITE;
+    ev.data.fd = fd;
+    epoll_ctl(EvFd(), EPOLL_CTL_MOD, fd, &ev);
+  } else {  // If it is a write multiplex, add the event
+    struct epoll_event ev {};
+    ev.events = EVENT_WRITE;
+    ev.data.fd = fd;
+    epoll_ctl(EvFd(), EPOLL_CTL_ADD, fd, &ev);
+  }
+}
```

Comment:
```suggestion
    struct epoll_event ev {};
    ev.events = EVENT_WRITE;
    ev.data.fd = fd;
    epoll_ctl(EvFd(), EPOLL_CTL_MOD, fd, &ev);
  } else {  // If it is a write multiplex, add the event
    struct epoll_event ev {};
    ev.events = EVENT_WRITE;
    ev.data.fd = fd;
    epoll_ctl(EvFd(), EPOLL_CTL_ADD, fd, &ev);
  }
}
```

这里可以用一个宏进行代码精简
```C++
#define EPOLL_EVENT(event_mode, evnet_ctl_mode)
do {
 struct epoll_event ev{};
ev.events = event_mode;
ev.data.fd = fd;
epoll_ctl(EvFd(), event_ctl_mode, fd, &v)
} while(0)
```

用 do...while(0) 主要是为了防止宏解析出错